twoliter/src/lock.rs

```diff
@@ -27,4 +37,208 @@ pub(crate) struct LockedImage {
+    #[serde(skip)]
+    is_sdk: bool,
+}
+
+/// The hash should not contain the source to allow for collision detection
+impl Hash for LockedImage {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.name.hash(state);
+        self.version.hash(state);
+        self.vendor.hash(state);
+    }
+}
+
+#[derive(Deserialize, Debug)]
+#[allow(dead_code)]
+struct ImageMetadata {
+    /// The name of the kit
+    pub name: String,
+    /// The version of the kit
+    pub version: Version,
+    /// The required sdk of the kit,
+    pub sdk: Image,
+    /// Any dependent kits
+    pub kit: Vec<Image>,
+}
+
+#[derive(Deserialize, Debug)]
+#[allow(dead_code)]
+struct ManifestListView {
+    manifests: Vec<ManifestView>,
+}
+
+#[derive(Deserialize, Debug)]
+#[allow(dead_code)]
+struct ManifestView {
+    digest: String,
+}
+
+macro_rules! docker {
+    ($arg: expr, $error_msg: literal, $($fmt: expr),*) => {{
+        let output = Command::new("docker")
+            .args($arg)
+            .output()
+            .await
+            .context(format!($error_msg, $($fmt),*))?;
+        ensure!(output.status.success(), $error_msg, $($fmt),*);
+        output.stdout
+    }};
+}
+
+#[allow(dead_code)]
+impl Lock {
+    pub(crate) async fn load(project: &Project) -> Result<Self> {
+        let lock_file_path = project.project_dir().join("Twoliter.lock");
+        if lock_file_path.exists() {
+            let lock_str = read_to_string(&lock_file_path)
+                .await
+                .context("failed to read lockfile")?;
+            let lock: Self =
+                toml::from_str(lock_str.as_str()).context("failed to deserialize lockfile")?;
+            // The digests must match, if changes are needed twoliter
+            ensure!(lock.digest == project.digest()?, "changes have occured to Twoliter.toml that require an update to Twoliter.lock, if intentional please run twoliter update");
+            return Ok(lock);
+        }
+        Self::create(project).await
+    }
+
+    pub(crate) async fn create(project: &Project) -> Result<Self> {
+        let lock_file_path = project.project_dir().join("Twoliter.lock");
+        if lock_file_path.exists() {
+            remove_file(&lock_file_path).await?;
+        }
+        let lock = Self::resolve(project).await?;
+        let lock_str = toml::to_string(&lock).context("failed to serialize lock file")?;
+        write(&lock_file_path, lock_str)
+            .await
+            .context("failed to write lock file")?;
+        Ok(lock)
+    }
+
+    async fn resolve(project: &Project) -> Result<Self> {
+        let vendor_table = project.vendor();
+        let mut known: HashMap<(String, String), Version> = HashMap::new();
+        let mut locked: Vec<LockedImage> = Vec::new();
+
+        let mut resolve: Vec<Image> = project.kits();
+        if let Some(sdk) = project.sdk_image() {
+            let mut sdk = sdk.clone();
+            sdk.is_sdk = true;
+            resolve.push(sdk);
+        }
```

Comment:
Rather than adding an `is_sdk` field to image that's only used for this algorithm, you can just maintain a set of SDKs that you find during traversal:
```suggestion
        let mut sdks = HashSet::new();
        if let Some(sdk) = project.sdk_image() {
            sdks.push(sdk.clone());
            resolve.push(sdk.clone());
        }
```

(need to derive Hash on Image but that should be trivial)